api_tests/harness.js

```diff
@@ -16,202 +25,272 @@ const project_root = execSync("git rev-parse --show-toplevel", {
-        ],
-    });
-}
+    stopLedgers() {
+        let names = Object.keys(this.running_ledgers);
+        if (names.length > 0) {
+            console.log("Stopping ledgers: " + names.join(", "));
 
-async function generateBlock(ledgers) {
-    if (ledgers && ledgers.includes("bitcoin")) {
-        await bitcoin.btc_generate();
+            Object.values(this.block_timers).forEach(clearInterval);
+
+            execSync("docker-compose rm -sfv", docker_compose_options);
+        }
+        this.running_ledgers = {};
     }
 }
 
-async function startComitNode(name, comit_config) {
-    return await spawn(project_root + "/target/debug/comit_node", [], {
-        cwd: services_cwd,
-        encoding: "utf-8",
-        env: { COMIT_NODE_CONFIG_PATH: comit_config.config_dir },
-        stdio: [
-            "ignore",
-            fs.openSync(log_dir + "/comit_node-" + name + ".log", "w"),
-            fs.openSync(log_dir + "/comit_node-" + name + ".log", "w"),
-        ],
-    });
-}
+class ComitRunner {
+    constructor() {
+        this.running_nodes = {};
+    }
 
-async function startBtsieve(name, btsieve_config) {
-    return await spawn(project_root + "/target/debug/btsieve", [], {
-        cwd: services_cwd,
-        encoding: "utf-8",
-        env: btsieve_config.env,
-        stdio: [
-            "ignore",
-            fs.openSync(log_dir + "/btsieve-" + name + ".log", "w"),
-            fs.openSync(log_dir + "/btsieve-" + name + ".log", "w"),
-        ],
-    });
-}
+    async ensureComitNodesRunning(comit_nodes) {
+        console.log(
+            "Starting comit node for " +
+                comit_nodes.map(([name, _]) => name).join(", ")
+        );
+        for (let [name, comit_config] of comit_nodes) {
+            if (this.running_nodes[name]) {
+                continue;
+            }
 
-// ********************************** //
-// Start services, run test, shutdown //
-// ********************************** //
-
-function run_tests(file) {
-    describe("🏃" + file, async function() {
-        let block_interval;
-        let subprocesses = [];
-        before(async function() {
-            this.timeout(ledger_up_time + 4000);
-
-            if (config.ledgers) {
-                const subprocess = await startDockerContainers(
-                    config.ledgers,
-                    ledgers_config
+            this.running_nodes[name] = await spawn(
+                project_root + "/target/debug/comit_node",
+                [],
+                {
+                    cwd: test_root,
+                    encoding: "utf-8",
+                    env: { COMIT_NODE_CONFIG_PATH: comit_config.config_dir },
+                    stdio: [
+                        "ignore",
+                        fs.openSync(
+                            log_dir + "/comit_node-" + name + ".log",
+                            "w"
+                        ),
+                        fs.openSync(
```

Comment:
❓copy pasta or do you want to open the log file twice on purpose?

Reply:
Good question. I think it does work but it's not guaranteed to look pretty. When stderr and stdout are written at the same time then some lines might be a bit mangled.  We could write stdout and stderr to different log files but I think its better to keep them in one so we don't accidentally miss something.